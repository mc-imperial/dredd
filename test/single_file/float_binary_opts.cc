int main() {
  double x = 0.0 + 5.32;
  double y = 5.234 + 2.352;
}
