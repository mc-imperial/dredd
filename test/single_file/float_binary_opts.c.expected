#include <inttypes.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#ifdef _MSC_VER
#define thread_local __declspec(thread)
#elif __APPLE__
#define thread_local __thread
#else
#include <threads.h>
#endif

static thread_local int __dredd_some_mutation_enabled = 1;
static int __dredd_enabled_mutation(int local_mutation_id) {
  static thread_local int initialized = 0;
  static thread_local uint64_t enabled_bitset[1];
  if (!initialized) {
    int some_mutation_enabled = 0;
    const char* dredd_environment_variable = getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable) {
      char* temp = malloc(strlen(dredd_environment_variable) + 1);
      strcpy(temp, dredd_environment_variable);
      char* token;
      token = strtok(temp, ",");
      while(token) {
        int value = atoi(token);
        int local_value = value - 0;
        if (local_value >= 0 && local_value < 29) {
          enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
          some_mutation_enabled = 1;
        }
        token = strtok(NULL, ",");
      }
      free(temp);
    }
    initialized = 1;
    __dredd_some_mutation_enabled = some_mutation_enabled;
  }
  return enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64));
}

static double __dredd_replace_expr_double_zero(double arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return 1.0;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return -1.0;
  return arg;
}

static double __dredd_replace_expr_double(double arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return -(arg);
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return 0.0;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return 1.0;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return -1.0;
  return arg;
}

static double __dredd_replace_binary_operator_Add_arg1_double_arg2_double_lhs_zero(double arg1, double arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1 + arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1 - arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg2;
  return arg1 + arg2;
}

static double __dredd_replace_binary_operator_Add_arg1_double_arg2_double(double arg1, double arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1 + arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1 / arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg1 * arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg1 - arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg1;
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return arg2;
  return arg1 + arg2;
}

int main() {
  double x = __dredd_replace_expr_double(__dredd_replace_binary_operator_Add_arg1_double_arg2_double_lhs_zero(__dredd_replace_expr_double_zero(0.0, 0) , __dredd_replace_expr_double(5.32, 2), 6), 8);
  double y = __dredd_replace_expr_double(__dredd_replace_binary_operator_Add_arg1_double_arg2_double(__dredd_replace_expr_double(5.234, 12) , __dredd_replace_expr_double(2.352, 16), 20), 25);
}
